File: cpp/leetcode/377.cpp
int combinationSum4(vector<int>& nums, int target) {
    vector<int> res(target + 1);
    res[0] = 1;
    for (int i = 1; i <= target; ++i) {
        for (int x : nums) {
            if (i >= x) res[i] += res[i - x];
        }
    }
    return res[target];
}

File: python/leetcode/377.py
class Solution(object):
    def combinationSum4(self, nums, target):
        nums, combs = sorted(nums), [1] + [0] * target
        for i in range(target + 1):
            for num in nums:
                if num > i: break
                if num == i: combs[i] += 1
                if num < i: combs[i] += combs[i - num]
        return combs[target]

class Solution(object):
    def combinationSum4WithLength(self, nums, target, length, memo=collections.defaultdict(int)):
        if length <= 0: return 0
        if length == 1: return 1 * (target in nums)
        if (target, length) not in memo:
            for num in nums:
                memo[target, length] += self.combinationSum4(nums, target - num, length - 1)
        return memo[target, length]